core/lib/src/local/form.rs

```diff
@@ -0,0 +1,222 @@
+use crate::{
+    form::{name::NameBuf, ValueField},
+    http::uri::Uri,
+};
+use rocket_http::{ContentType, RawStr};
+
+/// A struct that can be used to easily create simple and multipart form data
+/// for testing purposes.
+///
+/// ## Example
+///
+/// The following snippet creates a request with a form submission:
+///
+/// ```rust,no_run
+/// use rocket::local::form::LocalForm;
+/// use rocket::local::blocking::{Client, LocalRequest};
+/// use rocket::http::{ContentType};
+///
+/// let client = Client::tracked(rocket::ignite()).expect("valid rocket");
+/// let req = client.post("/")
+///     .form(LocalForm::new()
+///             .field("field", "value")
+///             .file("foo.txt", ContentType::Plain, "hi there!"));
+///
+/// let response = req.dispatch();
+/// ```
+pub struct LocalForm<'v>(Vec<LocalField<'v>>);
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum LocalField<'v> {
+    Value(NameBuf<'v>, String),
+    Data(
+        NameBuf<'v>,
+        Option<&'v str>,
+        ContentType,
+        Vec<u8>,
+    ),
+}
+
+impl<'v> LocalForm<'v> {
+    pub fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    /// A percent-decoded `name` and `value`.
+    pub fn field<N, V>(mut self, name: N, value: V) -> Self
+    where
+        N: Into<NameBuf<'v>>,
+        V: AsRef<str>,
+    {
+        self.0
+            .push(LocalField::Value(name.into(), value.as_ref().to_owned()));
+        self
+    }
+
+    /// Adds all the `fields` to `self`.
+    pub fn fields<I, F>(mut self, fields: I) -> Self
+    where
+        I: Iterator<Item = F>,
+        F: Into<ValueField<'v>>,
+    {
+        fields.for_each(|field| {
+            let field = field.into();
+            self.0.push(LocalField::Value(
+                NameBuf::from(field.name.as_name()),
+                field.value.to_string(),
+            ));
+        });
+        self
+    }
+
+    /// A percent-encoded `name` and `value`.
+    pub fn raw_field(mut self, name: &'v RawStr, value: &'v RawStr) -> Self {
+        let decoded_name = name.percent_decode_lossy().into_owned();
+        let value = value.percent_decode_lossy().into_owned();
+        self.0
+            .push(LocalField::Value(NameBuf::from(decoded_name), value));
+        self
+    }
+
+    /// Adds a field for the file with the name `file_name`, content-type `ct`, and
+    /// file contents `data`.
+    pub fn file<N, V>(mut self, file_name: N, ct: ContentType, data: V) -> Self
+    where
+        N: Into<Option<&'v str>>,
+        V: AsRef<[u8]>,
+    {
+        self.0.push(LocalField::Data(
+            NameBuf::from("file"),
```

Comment:
I didn't have it in my original API, but this field name needs to be a parameter to the function. We can't just set it to `file`.